### PR TITLE
Add a warning/additional handlers for parsing`synthetic-package`.

### LINKED ISSUE #157928: Flutter i10n generator should warn if a synthetic package (`flutter_gen`) is used
See deprecation notice: https://github.com/flutter/website/pull/11343.

The warning should link to the migration guide.

### LINKED ISSUE #157929: Flutter i10n generator should disallow synthetic packages with `--no-implicit-pubspec-resolution`.
I landed an initial version of this in https://github.com/flutter/flutter/issues/157819, but I think we should make a slight change - instead of just using `implicit-pubspec-resolution` as the default value for `synthetic-package`, if `--no-implicit-pubspec-resolution` is set, it should be an error (`throwToolExit`) to use a synthetic package if the flag is disabled.

### PR DESCRIPTION
**Pull request by matanlurey**
Closes https://github.com/flutter/flutter/issues/157928.
Closes https://github.com/flutter/flutter/issues/157929.

| Condition | Expectation |
| --------- | ------------ |
| `synthetic-packages: true` && `--implicit-pubpsec-resolution` | Generates `flutter_gen` with warning.
| `<no synthetic-packages key>` && `--implicit-pubspec-resolution` | Generates `flutter_gen` with warning.
| `synthetic-packages: false` && `--implicit-pubpsec-resolution` | Does not generate `flutter_gen`.
| `synthetic-packages: true` && `--no-implicit-pubpsec-resolution` | Error.
| `<no synthetic-packages key>` && `--no-implicit-pubspec-resolution` | Does not generate `flutter_gen`.
| `synthetic-packages: false` && `--no-implicit-pubpsec-resolution` | Generates `flutter_gen` with warning.

### CODE REVIEW

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

File: packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart:57]: Typo?

```suggestion
      'Cannot generate a synthetic package when --no-implicit-pubspec-resolution is passed.\n'
```

File: packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart:71]: Rhetorical question: what is "this value"?

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Review by andrewkolos — CHANGES_REQUESTED** [review]
I have some things I need to intensely review here, but here is some stuff to get started in case that's useful. Otherwise, simply wait for another review from me.

File: packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart
Comment (matanlurey) [reply packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart]: Done.

File: packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart
Comment (matanlurey) [reply packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart]: No, but I made it look nicer.

File: packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart
Comment (matanlurey) [reply packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart]: No, but I made it look nicer.

File: packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart
Comment (matanlurey) [reply packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart]: Clarified.

File: packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/dart/generate_synthetic_packages.dart:76]:
```suggestion
  environment.logger.printWarning(
    'Synthetic package output (package:flutter_gen) is deprecated: '
    'https://flutter.dev/to/flutter-gen-deprecation. In a future release, '
    'synthetic-package will default to `false` and will later be removed '
    'entirely.',
  );
```

**Review by andrewkolos — APPROVED** [review]
LGTM

**Comment by polina-c** [comment]
reason for revert: b/377107864

**Pull request merged**
